[PROTOCOL-407] Track loans with userId

tpscrpt opened the pull request:
userId is a field provided in the loanRequest and verified through the standard loanTermsConsensus hashing/signature verification mechanism. It is generated in the centralized data-provider based on the user's identifiable details.

We ensure that a provided userId is not used to take out more than 1 unsecured loan (< 30% collateral).

We do this by mapping userIds to a boolean (could be changed to a limit on number of loans or total lent in the future) and requiring that this value be false for a given loanRequest in withValidLoanRequest.

The value is set to `true` for a given userId at the end of createLoanWithTerms and back to `false` when repay leaves a loan with no outstanding balance.

EDIT: if a user doesn't provide bank info on the validator side, the userId is set to bytes32(0).

passabilities commented on contracts/base/Escrow.sol:
```diff
@@ -153,7 +153,10 @@ contract Escrow is EscrowInterface, TInitializable, BaseEscrowDapp {
     function claimTokens() external onlyOwner() {
-        require(getLoan().status == TellerCommon.LoanStatus.Closed, "LOAN_NOT_CLOSED");
+        require(
+            getLoan().status == TellerCommon.LoanStatus.Liquidated,
```
Because this function can only be called by the owner, we should not be checking that the loan is `Liquidated` but rather just `Closed`. If a user pays back their loan, we still want them to be able to claim their tokens that are still in the Escrow.

passabilities commented on test/base/EscrowClaimTokensTest.js:
```diff
@@ -31,24 +31,23 @@ contract("EscrowClaimTokensTest", function(accounts) {
-    _3_loan_not_liquidated_recipient_is_borrower: [ loanStatus.Closed, accounts[1], accounts[1], false, 2, 1000, true, "LOAN_NOT_LIQUIDATED" ],
-    _4_loan_liquidated_recipient_is_borrower: [ loanStatus.Closed, accounts[1], accounts[1], true, 2, 1000, false, null ],
-    _5_loan_liquidated_recipient_not_borrower: [ loanStatus.Closed, accounts[1], accounts[2], true, 2, 1000, true, "CALLER_MUST_BE_LOANS" ],
+    _1_loan_active: [ loanStatus.Active, accounts[1], accounts[1], 0, 0, true, "LOAN_NOT_LIQUIDATED" ],
+    _2_loan_not_liquidated_recipient_not_borrower: [ loanStatus.Closed, accounts[1], accounts[2], 0, 0, true, "LOAN_NOT_LIQUIDATED" ],
+    _3_loan_not_liquidated_recipient_is_borrower: [ loanStatus.Closed, accounts[1], accounts[1], 2, 1000, true, "LOAN_NOT_LIQUIDATED" ],
+    _4_loan_liquidated_recipient_is_borrower: [ loanStatus.Liquidated, accounts[1], accounts[1], 2, 1000, false, null ],
+    _5_loan_liquidated_recipient_not_borrower: [ loanStatus.Liquidated, accounts[1], accounts[2], 2, 1000, true, "CALLER_MUST_BE_LOANS" ],
```
No need to be checking that the loan is `Liquidated` 

1. Should throw when loan is active
1. Should throw when caller not the owner
1. Should only pass when loan is `Closed` and caller is owner

passabilities commented on test/base/EstimateGasEscrowClaimTokensTest.js:
```diff
@@ -17,8 +17,8 @@ const Escrow = artifacts.require("./mock/base/EscrowMock.sol");
-  const baseGasCost = 452022; // Gas cost with 1 token in wallet
-  const expectedGasCost = (tokens) => baseGasCost + ((tokens -  1) * 37000); // Gas cost > 1 token in wallet
+  const baseGasCost = 453453; // Gas cost with 1 token in wallet
+  const expectedGasCost = (tokens) => baseGasCost + ((tokens -  1) * 167151); // Gas cost > 1 token in wallet
```
this is quite a hike in expected gas cost @tpscrpt @sypeer

tpscrpt replied:
Yeah I'll review this again and try to find why, currently my best guess is just the added bytes(32) read from storage on every getLoan() call (since it's kept in another contract and can't just extract 1 value). Might have to refactor so this loan gets passed around instead of read multiple times from storage.

passabilities commented on test/base/EstimateGasEscrowClaimTokensTest.js:
```diff
@@ -48,7 +48,7 @@ contract("EstimateGasEscrowClaimTokensTest", function(accounts) {
       await loans.givenMethodReturn(
         loansEncoder.encodeLoans(),
-        encodeLoanParameter(web3, { status: Closed, liquidated: true, loanTerms: { borrower } })
+        encodeLoanParameter(web3, { status: Liquidated, loanTerms: { borrower } })
```
Should just be `Closed`

passabilities commented on test/base/EtherCollateralLoansGetCollateralInfoTest.js:
```diff
@@ -120,6 +121,8 @@ contract("EtherCollateralLoansGetCollateralInfoTest", function(accounts) {
+        case LIQUIDATED:
+          collateralNeededInTokens = new BigNumber(0);
         case CLOSED:
           collateralNeededInTokens = new BigNumber(0);
```
Can remove line 125 and just let the switch statement fall through

passabilities commented on scripts/utils/forkRopsten.js:
```diff
@@ -0,0 +1,46 @@
+const Ganache = require("ganache-core");
```
Lets add a CLI param to pass in the network flag for the existing `forkMainnet.js`

passabilities commented on contracts/base/LoansBase.sol:
```diff
@@ -241,6 +247,10 @@ contract LoansBase is LoansInterface, Base {
 
+        if (request.userId != bytes32(0)) {
+            userIdHasLoan[request.userId] = true;
+        }
```
We'll need to validate that this flag is not already set OR if it is check their previous unsecured loan status. If its `TermsSet`, meaning it was never taken out, validate that the terms have expired.

Also I do not see a check on the `collateralRatio` value @tpscrpt this should only happen when the collateral ratio is 0, right?

passabilities commented on contracts/base/LoansBase.sol:
```diff
@@ -420,6 +430,11 @@ contract LoansBase is LoansInterface, Base {
+            bytes32 userId = loans[loanID].userId;
+            if (userId != bytes32(0)) {
+                userIdHasLoan[userId] = false;
+            }
```
We'll need to put this when a loan is liquidated as well since the loan is also repaid their.

tpscrpt commented:
@passabilities https://github.com/teller-protocol/teller-protocol-v1/pull/232/commits/f7333988263dee5332be687435d417060ad70838

Addresses some concerns with TermsSet loans, read the commit message and code diff for context

passabilities commented on contracts/base/LoansBase.sol:
```diff
@@ -383,6 +396,10 @@ contract LoansBase is LoansInterface, Base {
             EscrowInterface(loans[loanID].escrow).initialize(address(this), loanID);
         }
 
+        if (loans[loanID].userId != bytes32(0)) {
+            userIdHasLoan[loans[loanID].userId] = true;
+        }
```
I assume that because the `loans[loanID].userId` value is set it meas that the loan is unsecured? Lets also add the check `!loans[loanID].isLoanSecured()` which is the solidity implementation.

passabilities commented on contracts/base/LoansBase.sol:
```diff
@@ -626,4 +646,15 @@ contract LoansBase is LoansInterface, Base {
+        @param loanID loan id associated to the Escrow contract.
+     */
+    function _clearUserIdLoan(uint256 loanID) internal {
+        bytes32 userId = loans[loanID].userId;
```
No need for local variable here; just adds more gas cost

passabilities commented on contracts/base/LoansBase.sol:
```diff
@@ -626,4 +646,15 @@ contract LoansBase is LoansInterface, Base {
+     */
+    function _clearUserIdLoan(uint256 loanID) internal {
+        bytes32 userId = loans[loanID].userId;
+        if (userId != bytes32(0)) {
```
Also we need to be checking that the `loanID` here is unsecured (`!loans[loanID].isSecured()`) before we clear the boolean
